modules/server/src/main/scala/scaladex/server/service/SonatypeSynchronizer.scala

```diff
@@ -0,0 +1,73 @@
+package scaladex.server.service
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+import com.typesafe.scalalogging.LazyLogging
+import scaladex.core.model.Artifact._
+import scaladex.core.service.SchedulerDatabase
+import scaladex.core.service.SonatypeService
+import scaladex.core.util.ScalaExtensions._
+
+class SonatypeSynchronizer(
+    database: SchedulerDatabase,
+    sonatypeService: SonatypeService,
+    publishProcess: PublishProcess
+)(implicit ec: ExecutionContext)
+    extends Scheduler("sync-sonatype", 24.hours)
+    with LazyLogging {
+  import SonatypeSynchronizer._
+  override def run(): Future[Unit] =
+    for {
+      groupIds <- database.getAllGroupIds()
+      mavenReferenceFromDatabase <- database.getAllMavenReferences()
+      // we sort just to estimate through the logs the percentage of progress
+      result <- groupIds.sortBy(_.value).mapSync(g => findAndIndexMissingArtifacts(g, mavenReferenceFromDatabase))
+      _ = logger.info(s"${result.size} poms have been successfully indexed")
+    } yield ()
+
+  def findAndIndexMissingArtifacts(groupId: GroupId, mavenReferenceFromDatabase: Seq[MavenReference]): Future[Int] =
+    sonatypeService
+      .getAllArtifactIds(groupId)
+      .flatMap { artifactIds =>
+        artifactIds.grouped(6).mapSync { artifactIds =>
+          artifactIds.map(id => findAndIndexMissingArtifacts(groupId, id, mavenReferenceFromDatabase.toSet)).sequence
+        }
+      }
+      .map(_.flatten.sum)
+
+  private def findAndIndexMissingArtifacts(
+      groupId: GroupId,
+      artifactId: ArtifactId,
+      mavenReferenceFromDatabase: Set[MavenReference]
+  ): Future[Int] =
+    if (artifactId.isScala()) {
+      for {
+        versions <- sonatypeService.getAllVersions(groupId, artifactId)
+        mavenReferences = versions.map(v =>
+          MavenReference(groupId = groupId.value, artifactId = artifactId.value, version = v.toString)
+        )
+        missingVersions = findMissingVersions(mavenReferenceFromDatabase, mavenReferences)
+        _ = if (missingVersions.nonEmpty)
+          logger.warn(s"${missingVersions.size} artifacts are missing for ${groupId.value}:${artifactId.value}")
+        else ()
+        missingPomFiles <- missingVersions
+          .mapSync(ref => sonatypeService.getPomFile(ref).map(_.map(ref -> _)))
+          .map(_.flatten)
+        publishResult <- missingPomFiles.mapSync {
+          case (mavenRef, (pomFile, creationDate)) =>
+            publishProcess.publishPom(s"${mavenRef.nameWithVersion}", pomFile, creationDate, None)
+        }
+      } yield publishResult.count {
+        case PublishResult.Success => true
+        case _                     => false
+      }
+    } else Future.successful(0)
```

Comment:
If a function does not do anything in some case, I would prefer not to call it. Meaning we should filter scala artifacts before calling this fonction and not test here if it is a scala artifact or not.

Reply:
ok